x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/MultiMatch.java

```diff
@@ -0,0 +1,454 @@
+    }
+
+    private MultiMatch(Source source, Expression query, List<Expression> fields, Expression options, QueryBuilder queryBuilder) {
+        super(source, query, initChildren(query, fields, options), queryBuilder);
+        this.options = options;
+        this.fields = fields;
+    }
+
+    private static MultiMatch readFrom(StreamInput in) throws IOException {
+        Source source = Source.readFrom((PlanStreamInput) in);
+        Expression query = in.readNamedWriteable(Expression.class);
+        List<Expression> fields = in.readNamedWriteableCollectionAsList(Expression.class);
+        QueryBuilder queryBuilder = null;
+        if (in.getTransportVersion().onOrAfter(TransportVersions.ESQL_QUERY_BUILDER_IN_SEARCH_FUNCTIONS)) {
+            queryBuilder = in.readOptionalNamedWriteable(QueryBuilder.class);
+        }
+        return new MultiMatch(source, query, fields, null, queryBuilder);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(query());
+        out.writeNamedWriteableCollection(fields);
+        if (out.getTransportVersion().onOrAfter(TransportVersions.ESQL_QUERY_BUILDER_IN_SEARCH_FUNCTIONS)) {
+            out.writeOptionalNamedWriteable(queryBuilder());
+        }
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public Expression replaceChildren(List<Expression> newChildren) {
+        return new MultiMatch(
+            source(),
+            newChildren.getFirst(),
+            newChildren.subList(1, newChildren.size() - 1),
+            newChildren.getLast(),
+            queryBuilder()
+        );
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, MultiMatch::new, query(), fields, options);
+    }
+
+    @Override
+    protected Query translate(TranslatorHandler handler) {
+        Map<String, Float> fieldsWithBoost = new HashMap<>();
+        for (Expression field : fields) {
+            var fieldAttribute = Match.fieldAsFieldAttribute(field);
+            Check.notNull(fieldAttribute, "MultiMatch must have field attributes as arguments #2 to #N-1.");
+            String fieldName = Match.getNameFromFieldAttribute(fieldAttribute);
+            fieldsWithBoost.put(fieldName, 1.0f);
+        }
+        return new MultiMatchQuery(source(), Objects.toString(queryAsObject()), fieldsWithBoost, getOptions());
+    }
+
+    @Override
+    public Expression replaceQueryBuilder(QueryBuilder queryBuilder) {
+        return new MultiMatch(source(), query(), fields, options, queryBuilder);
+    }
+
+    public List<Expression> fields() {
+        return fields;
+    }
+
+    public Expression options() {
+        return options;
+    }
+
+    private Map<String, Object> getOptions() throws InvalidArgumentException {
+        if (options() == null) {
```

Comment:
I think we should add `lenient`: `true` as an option by default, similar to match. That helps when users use a field that has different types on different indices, as otherwise the query will fail. See how `Match` implements it.